Testing System/src/main/java/com/epam/practice/testingsystem/data/dao/DataParse.java

```diff
@@ -0,0 +1,158 @@
+import java.util.List;
+import java.util.function.Function;
+
+class DataParse {
+    static User getUser(ResultSet rs) {
+        if (rs == null)
+            throw new IllegalArgumentException();
+        try {
+            int id = rs.getInt("id");
+            String name = rs.getString("user_name");
+            String pass = rs.getString("password_hash");
+            int roleId = rs.getInt("role_id");
+            return new User(id, name, pass, roleId);
+        }
+        catch (SQLException e) {
+            e.printStackTrace();
+            throw new RuntimeException("Incorrect data structure");
+        }
+    }
+
+    static UniversityGroup getUniversityGroup(ResultSet rs) {
+        if (rs == null)
+            throw new IllegalArgumentException();
+        try {
+            UniversityGroup universityGroup = new UniversityGroup();
+            int id = rs.getInt("id");
+            String name = rs.getString("name");
+            universityGroup.setId(id);
+            universityGroup.setName(name);
+            return universityGroup;
+        }
+        catch (SQLException e) {
+            e.printStackTrace();
+            throw new RuntimeException("Incorrect data structure");
+        }
+    }
+
+    static Answer getAnswer(ResultSet rs) {
+        if (rs == null)
+            throw new IllegalArgumentException();
+        try {
+            int id = rs.getInt("id");
+            String answer = rs.getString("answer");
+            Boolean isCorrect = rs.getBoolean("is_correct");
+            if (rs.wasNull())
+                isCorrect = null;
+            return new Answer(id, answer, isCorrect);
+        }
+        catch (SQLException e) {
+            e.printStackTrace();
+            throw new RuntimeException("Incorrect data structure");
+        }
+    }
+
+    static Question getQuestion(ResultSet rs) {
+        if (rs == null)
+            throw new IllegalArgumentException();
+        try {
+            int id = rs.getInt("id");
+            String question = rs.getString("question");
+            int score = rs.getInt("score");
+            List<Answer> answers = new AnswerDAO().findAll(id);
+            return new Question(id, question, score, answers);
+        }
+        catch (SQLException e) {
+            e.printStackTrace();
+            throw new RuntimeException("Incorrect data structure");
+        }
+    }
+
+    static Attempt getAttempt(ResultSet rs) {
+        if (rs == null)
+            throw new IllegalArgumentException();
+        try {
+            Attempt attempt = new Attempt();
+            int id = rs.getInt("id");
+            int testId = rs.getInt("test_id");
+            int userId = rs.getInt("user_id");
+            int score = rs.getInt("score");
+            Time datetime = rs.getTime("datetime");
+            attempt.setId(id);
+            attempt.setDateTime(datetime);
+            attempt.setScore(score);
+
+            ITestDAO testDAO = new TestDAO();
```

Comment:
This part could go to another private function. The whole method consists of different parts and it would be good to know which part has failed if something went wrong.

One possible way to do it is to separate the logic and move it into another small function and write unit tests for them. This way you will be sure your codebase is reliable.

Reply:
DAO object creation was replaced with calling `DAOFactory`